### DISCUSSION TITLE
using jinja extends to create a single file

### DISCUSSION THREAD

**Discussion by awesomescot** [copier-org/copier#1465] [Q&A] [answered]
Hi.  Thank you for creating this wonderful tool.  I've done my best to search for this question, but haven't found anything.  Here's what I'm looking for.

I would like to create a single file with a default configuration(eg. config.yaml), but also enable modifications to this file for specific apps, which there are many.  I'd like to avoid massive logic in the default config template for each app.  So I'd like to have a directory of files for the specific apps with a file for each app that would modify the default.  (eg. /config/<app_name>.yaml.jinja.  

config.yaml.jinja ->(pulls aditional config from if it exists)->config/<app_name>.yaml.jinja -> to create a single file config.yaml

It seems like the [extends](https://jinja.palletsprojects.com/en/3.1.x/templates/#template-inheritance) inheritance model using blocks in jinja would work perfectly for this.  But I think I need some python code to run which does some logic like this?

`generate_template_from('config/my_app.yaml.jinja') if os.path.isfile('config/my_app.yaml.jinja') else generate_template_from('config.yaml.jinja')` 

Does anyone have any ideas for how to implement this?

**Comment by yajo** [comment]
Welcome to copier! I'm glad you like it 😊 

I'm having a bit of trouble to understand the case you're describing. Could you post some example of what you've done and what would you want to avoid please? 

Thanks!

**Reply by awesomescot** [reply]
Ah, sorry for my messy explanation.  Thanks for taking a look.  I got the answer below.

**Comment by sisp** [comment, marked as answer]
I think you're looking for Jinja inheritance and a Jinja include with a dynamic filename. Perhaps like this?

* `includes/config/_base.yaml.jinja` contains your base config, perhaps with some Jinja blocks as extension hooks.
* `includes/config/<app_name>.yaml.jinja` extend `_base.yaml.jinja`, e.g.:
    ```yaml+jinja
    {% extends "includes/config/_base.yaml.jinja" %}
    # your app specific extension
    ```
* `template/config.yaml.jinja` includes the app specific config content:
    ```jinja
    {% include "includes/config/" ~ app_name ~ ".yaml.jinja" %}
    ```
    Here, I assume you set `_subdirectory: template/` in `copier.yml`, otherwise you need to make sure to exclude the `includes/` dir from being copied, and I assume you have a Copier question variable `app_name` (probably a choice question).

Does this help?

**Reply by awesomescot** [reply]
Yes, that's basically what I ended up with.  And my include used an array so if the app file doesn't exist it uses the base file.  
```
{% include ["includes/config/" ~ app_name ~ ".yaml.jinja", "includes/config/_base.yaml.jinja"] %}
```

It feels a bit awkward, but it works.  I would guess that this would be a common pattern?  Thanks for the help!!

**Reply by sisp** [reply]
I'm glad it works for you. :tada:

What exactly do you find awkward about it? I find it quite intuitive TBH. :laughing:

**Reply by awesomescot** [reply]
It's not so bad.  Maybe if this pattern was documented, that would be nice.  But now that I understand it, it makes sense.  Thanks!
